git-worktree/src/lib.rs

```diff
@@ -1 +1,111 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_index::{
+    entry::{Flags, Mode},
+    State,
+};
+use git_object::bstr::ByteSlice;
+use git_object::Data;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) { from() }
+        TimeError(err: std::time::SystemTimeError) { from() }
+        ToU32Error(err: std::num::TryFromIntError) { from() }
+        IoError(err: std::io::Error) { from() }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<Data<'a>>,
+{
+    let path = path.as_ref();
+    let mut buf = Vec::new();
+    let mut entry_time = Vec::new(); // Entries whose timestamps have to be updated
+    for (i, entry) in state.entries().iter().enumerate() {
+        if entry.flags.contains(Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let dest = path.join(entry.path(state).to_path()?);
+        create_dir_all(dest.parent().expect("path is empty"))?;
+        match entry.mode {
+            Mode::FILE | Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                std::fs::write(&dest, obj.data)?;
+                if entry.mode == Mode::FILE_EXECUTABLE {
+                    #[cfg(unix)]
+                    fs::set_permissions(&dest, fs::Permissions::from_mode(0o777)).unwrap();
+                }
+                let met = std::fs::metadata(&dest)?;
+                // Set both fields to mtime if ctime is not present and vice-versa
+                // If both fields are not present, will raise an Err
+                let ctime = met.created().or_else(|_| met.modified())?;
+                let mtime = met.modified().or_else(|_| met.created())?;
+                entry_time.push((ctime, mtime, i));
+            }
+            Mode::SYMLINK => {
+                let obj = find(&entry.id, &mut buf).unwrap();
+                let linked_to = obj.data.to_path()?;
+                if opts.symlinks {
+                    #[cfg(unix)]
+                    std::os::unix::fs::symlink(linked_to, &dest)?;
+                    #[cfg(windows)]
+                    if dest.exists() {
+                        if dest.is_file() {
+                            std::os::windows::fs::symlink_file(linked_to, &dest)?;
+                        } else {
+                            std::os::windows::fs::symlink_dir(linked_to, &dest)?;
+                        }
+                    }
+                } else {
+                    std::fs::write(&dest, obj.data)?;
+                }
+                let met = std::fs::metadata(&dest)?;
```

Comment:
For good measure, let's use `symlink_metadata()` everywhere. 
This is particularly important here as we are never interested in following the symlink we potentially just created to a file that definitely didn't change.